ledger/participant-integration-api/src/main/scala/platform/apiserver/services/admin/ApiParticipantPruningService.scala

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.platform.apiserver.services.admin
+
+import java.util.UUID
+
+import com.daml.dec.DirectExecutionContext
+import com.daml.ledger.api.v1.admin.participant_pruning_service.{
+  ParticipantPruningServiceGrpc,
+  PruneRequest
+}
+import com.daml.ledger.api.v1.ledger_offset.LedgerOffset
+import com.daml.ledger.participant.state.index.v2.{IndexParticipantPruningService, LedgerEndService}
+import com.daml.ledger.participant.state.v1.{
+  NotPruned,
+  Offset,
+  ParticipantPruned,
+  SubmissionId,
+  WriteParticipantPruningService
+}
+import com.daml.lf.data.Ref
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.platform.api.grpc.GrpcApiService
+import com.daml.platform.server.api.validation.ErrorFactories
+import com.google.protobuf.empty.Empty
+import io.grpc.{ServerServiceDefinition, StatusRuntimeException}
+
+import scala.compat.java8.FutureConverters
+import scala.concurrent.{ExecutionContext, Future}
+
+final class ApiParticipantPruningService private (
+    readBackend: IndexParticipantPruningService with LedgerEndService,
+    writeBackend: WriteParticipantPruningService)(
+    implicit grpcExecutionContext: ExecutionContext,
+    logCtx: LoggingContext)
+    extends ParticipantPruningServiceGrpc.ParticipantPruningService
+    with GrpcApiService {
+
+  private val logger = ContextualizedLogger.get(this.getClass)
+
+  override def bindService(): ServerServiceDefinition =
+    ParticipantPruningServiceGrpc.bindService(this, DirectExecutionContext)
+
+  override def prune(request: PruneRequest): Future[Empty] = {
+    val submissionIdOrErr = SubmissionId
+      .fromString(
+        if (request.submissionId.nonEmpty) request.submissionId else UUID.randomUUID().toString)
+      .left
+      .map(err => ErrorFactories.invalidArgument(s"submission_id $err"))
+
+    submissionIdOrErr.fold(
+      Future.failed,
+      submissionId =>
+        LoggingContext.withEnrichedLoggingContext("submissionId" -> submissionId) {
+          implicit logCtx =>
+            (for {
+
+              pruneUpTo <- validateRequest(request: PruneRequest)
+
+              _ <- pruneWriteService(pruneUpTo, submissionId)
+
+              pruneResponse <- pruneLedgerApiServerIndex(pruneUpTo)
```

Comment:
If the write service pruning succeeds but the index pruning fails, then the participant is probably left in a corrupt state.